v2/handler.js

```diff
@@ -0,0 +1,234 @@
+            });
+    }
+};
+
+TChannelV2Handler.prototype.handleInitRequest = function handleInitRequest(reqFrame) {
+    var self = this;
+    if (self.remoteHostPort !== null) {
+        self.emit('error', new Error('duplicate init request')); // TODO typed error
+        return;
+    }
+    var hostPort = reqFrame.body.hostPort;
+    var processName = reqFrame.body.processName;
+    self.remoteHostPort = hostPort;
+    // TODO: use processName
+    self.emit('identify.in', hostPort, processName);
+    self.sendInitResponse(reqFrame);
+};
+
+TChannelV2Handler.prototype.handleInitResponse = function handleInitResponse(resFrame) {
+    var self = this;
+    if (self.remoteHostPort !== null) {
+        self.emit('error', new Error('duplicate init response')); // TODO typed error
+        return;
+    }
+    var hostPort = resFrame.body.hostPort;
+    var processName = resFrame.body.processName;
+    // TODO: use processName
+    self.remoteHostPort = hostPort;
+    self.emit('identify.out', hostPort, processName);
+};
+
+TChannelV2Handler.prototype.handleCallRequest = function handleCallRequest(reqFrame) {
+    var self = this;
+    var id = reqFrame.id;
+    var name = String(reqFrame.body.arg1);
+
+    if (self.remoteHostPort === null) {
+        self.emit('error', new Error('call request before init request')); // TODO typed error
+        return;
+    }
+
+    var handler = self.channel.getEndpointHandler(name);
+    var headers = {};
+
+    // TODO: couple reqFrame.body.tracing for all downstream requests
+    // TODO: implement service name?
+
+    self.runInOp(handler, {
+        id: id,
+        arg1: reqFrame.body.arg1,
+        arg2: reqFrame.body.arg2,
+        arg3: reqFrame.body.arg3,
+        headers: headers
+    }, function sendResponseFrame(err, res1, res2, callback) {
+        self.sendResponseFrame(reqFrame, headers, err, res1, res2, callback);
+    });
+};
+
+TChannelV2Handler.prototype.handleCallResponse = function handleCallResponse(resFrame) {
+    var self = this;
+    if (self.remoteHostPort === null) {
+        self.emit('error', new Error('call response before init response')); // TODO typed error
+        return;
+    }
+    var id = resFrame.id;
+    var code = resFrame.body.code;
+    var arg1 = resFrame.body.arg1;
+    var arg2 = resFrame.body.arg2;
+    var arg3 = resFrame.body.arg3;
+    if (code === v2.CallResponse.Codes.OK) {
+        self.completeOutOp(null, id, arg2, arg3);
+    } else {
+        self.completeOutOp(TChannelApplicationError({
+            code: code,
+            arg1: arg1,
+            arg2: arg2,
+            arg3: arg3
+        }), id, arg2, null);
+    }
+};
+
+TChannelV2Handler.prototype.handleError = function handleError(errFrame) {
+    var self = this;
+    var id = errFrame.id;
+    var code = errFrame.body.code;
+    var message = errFrame.body.message;
+    var err = v2.ErrorResponse.CodeErrors[code]({message: message});
+    self.completeOutOp(err, id, null, null);
```

Comment:
this makes sense as an `err` because the error frame type is a protocol/tchannel error of the remote tchannel implementation.